5-appinfra/modules/cicd-pipeline/project-iam-bindings.tf

```diff
@@ -13,67 +13,107 @@
 # limitations under the License.
 
 locals {
-  cloud_build_sas = ["serviceAccount:${google_service_account.cloud_build.email}"] # cloud build service accounts used for CI
-  membership_re   = "projects/([^/]*)/locations/([^/]*)/memberships/([^/]*)$"
-  gke_projects = distinct(flatten([
-    for _, value in var.env_cluster_membership_ids : [
-      for item in value.cluster_membership_ids : regex(local.membership_re, item)[0]
-    ]
-  ]))
+  membership_re = "projects/([^/]*)/locations/([^/]*)/memberships/([^/]*)$"
+  envs          = keys(var.env_cluster_membership_ids)
+
+  memberships     = flatten([for i in local.envs : var.env_cluster_membership_ids[i].cluster_membership_ids])
+  memberships_map = { for i, item in local.memberships : (i) => item }
+  gke_projects    = { for i, item in local.memberships : (i) => regex(local.membership_re, item)[0] }
+}
+
+resource "google_project_iam_member" "cloud_trace_agent" {
+  project = var.project_id
+  role    = "roles/cloudtrace.agent"
+
+  member = data.google_compute_default_service_account.compute_service_identity.member
+}
+
+resource "google_project_iam_member" "metric_writer" {
+  project = var.project_id
+  role    = "roles/monitoring.metricWriter"
+
+  member = data.google_compute_default_service_account.compute_service_identity.member
+}
+
+resource "google_project_iam_member" "log_writer" {
+  for_each = {
+    "compute"      = data.google_compute_default_service_account.compute_service_identity.member,
+    "cloud_deploy" = google_service_account.cloud_deploy.member,
+    "cloud_build"  = google_service_account.cloud_build.member,
+  }
+  project = var.project_id
+  role    = "roles/logging.logWriter"
+
+  member = each.value
+}
+
+resource "google_project_iam_member" "builder" {
+  for_each = {
+    "cloud_build_service" = google_service_account.cloud_deploy.member,
+    "cloud_build"         = google_service_account.cloud_build.member,
+  }
+  project = var.project_id
+  role    = "roles/cloudbuild.builds.builder"
+
+  member = each.value
 }
 
-# authoritative project-iam-bindings to increase reproducibility
-module "project-iam-bindings" {
-  source   = "terraform-google-modules/iam/google//modules/projects_iam"
-  version  = "~> 8.0"
-  projects = [var.project_id]
-  mode     = "authoritative"
```

Comment:
Just curious if `"additive"` would have worked.

Reply:
Got a cycle error:

> Error: Invalid for_each argument
        	            	
        	            	  on .terraform/modules/cicd.project-iam-bindings/modules/projects_iam/main.tf line 52, in resource "google_project_iam_member" "project_iam_additive":
        	            	  52:   for_each = module.helper.set_additive
        	            	    ├────────────────
        	            	    │ module.helper.set_additive is a set of dynamic, known only after apply
        	            	
        	            	The "for_each" set includes values derived from resource attributes that
        	            	cannot be determined until apply, and so Terraform cannot determine the full
        	            	set of keys that will identify the instances of this resource.
        	            	
        	            	When working with unknown values in for_each, it's better to use a map value
        	            	where the keys are defined statically in your configuration and where only
        	            	the values contain apply-time results.
        	            	
        	            	Alternatively, you could use the -target planning option to first apply only
        	            	the resources that the for_each value depends on, and then apply a second
        	            	time to fully converge.

Reply:
Interesting

Reply:
https://console.cloud.google.com/cloud-build/builds/b4163550-4e82-43dd-b448-2ada7a0a6f15;step=24?project=cloud-foundation-cicd